File: src/lib/cpp/DSMySQLQuery.cpp
#include "DSMySQLQuery.h"
#include "DSCasts.h"

DynSoft::DSMySQLQuery::DSMySQLQuery(DSMySQLDatabase *db) : DSQuery(db) {
	db->Reconnect();
	connection = db->GetConnection();
	running = false;

	Prepare(_N("SET NAMES :encoding:"));
	Bind(_N("encoding"), db->GetEncoding(), false);
	Exec(false);
}

DynSoft::DSClassInfo DynSoft::DSMySQLQuery::GetClassInfo() {
	return DynSoft::DSClassInfo(
		DS_MYSQL_QUERY_NAME,
		DS_MYSQL_QUERY_VERSION,
		DS_MYSQL_QUERY_AUTHOR
	);
}

bool DynSoft::DSMySQLQuery::Exec(const bool storeResult) {
	bool ok = false;
	running = false;
	const char *tmpQuery = DSCasts::ToChar(GetParsedQuery());
	if(mysql_query(connection, tmpQuery) == 0) {
		if(!storeResult) {
			running = false;
			ok      = true;
		} else {
			result = mysql_store_result(connection);
			if(result != NULL) {
				running = true;
			} else {
				running = false;
			}

			ok = running;
		}
	}

	delete [] tmpQuery;
	return ok;
}

DynSoft::DSCasts DynSoft::DSMySQLQuery::GetField(const wxString &field, const bool caseSensitive) {
	return GetField(GetFieldNumber(field));
}

DynSoft::DSCasts DynSoft::DSMySQLQuery::GetField(const int column) {
	return row[column] == NULL ? DSCasts::Convert(wxString()) : DSCasts::Convert(row[column]);
}

bool DynSoft::DSMySQLQuery::IsNull(const wxString &field) {
	return IsNull(GetField(field).ToInt());
}

bool DynSoft::DSMySQLQuery::IsNull(const int column) {
	return row[column] == NULL;
}

DynSoft::DSCasts DynSoft::DSMySQLQuery::GetColumn(const int column) {
    MYSQL_FIELD *columns = mysql_fetch_fields(result);
	return DSCasts::Convert(columns[column].name);
}

unsigned int DynSoft::DSMySQLQuery::GetRowCount() const {
	if(running)
		return mysql_num_rows(result);
	return 0;
}

unsigned int DynSoft::DSMySQLQuery::GetColumnCount() const {
	return mysql_num_fields(result);
}

bool DynSoft::DSMySQLQuery::Begin() {
	return Query(_N("BEGIN"), false);
}

bool DynSoft::DSMySQLQuery::Commit() {
	return Query(_N("COMMIT"));
}

bool DynSoft::DSMySQLQuery::Rollback() {
	return Query(_N("ROLLBACK"));
}

void DynSoft::DSMySQLQuery::FreeResult() {
	running = false;
	mysql_free_result(result);
}

bool DynSoft::DSMySQLQuery::NextRecord() {
	if(running)
		if((row = mysql_fetch_row(result)) != NULL)
			return true;
	
	running = false;
	return false;
}

int DynSoft::DSMySQLQuery::GetFieldNumber(const wxString &field) const {
	int numColumns = mysql_num_fields(result);
    MYSQL_FIELD *columns = mysql_fetch_fields(result);
    for (int i = 0; i < numColumns; i++)
        if(field == DSCasts::ToString(columns[i].name))
            return i;
    return -1;
}

unsigned int DynSoft::DSMySQLQuery::GetLastInsertId() const {
	return (unsigned int) mysql_insert_id(connection);
}

File: src/lib/php/DSMySQLQuery.php
<?php

require_once "DSQuery.php";
require_once "DSMySQLDatabase.php";
require_once "DSClassInfo.php";

define("DS_MYSQL_QUERY_VERSION", "0.0.1");
define("DS_MYSQL_QUERY_NAME",    "DSMySQLQuery");
define("DS_MYSQL_QUERY_AUTHOR",  "Thorsten Geppert (thorsten@ext2fs.de)");

class DSMySQLQuery extends DSQuery {

	protected $connection = null;
	protected $running    = false;
	protected $result     = null;
	protected $row        = null;

	public function __construct(DSMySQLDatabase $db) {
		parent::__construct($db);
		
		$db->Reconnect();
		$this->connection = $db->GetConnection();
		$this->running    = false;
		
		$this->Prepare("SET NAMES :encoding:");
		$this->Bind("encoding", $db->GetEncoding(), false);
		$this->Exec(false);
	}

	public function GetClassInfo() {
		return new DSClassInfo(
			DS_MYSQL_QUERY_NAME,
			DS_MYSQL_QUERY_VERSION,
			DS_MYSQL_QUERY_AUTHOR
		);
	}

	public function Exec($storeResult = true) {
		$ok            = false;
		$this->running = false;
		$this->result = mysql_query($this->GetParsedQuery(), $this->connection);
		if($this->result) {
			if(!$storeResult) {
				$ok = true;
				$this->FreeResult();
			} else {
				$this->running = $this->result !== true;
				$ok            = $this->running;
			}
		}

		return $ok;
	}

	public function GetField($field) {
		return $this->row[$field];
	}

	public function IsNull($field) {
		return $this->GetField($field) == null;
	}

	public function GetColumn($column) {
		return mysql_field_name($this->result, $column);
	}

	public function GetRowCount() {
		if($this->running)
			return mysql_num_rows($this->result);
		return 0;
	}

	public function GetColumnCount() {
		return mysql_num_fields($this->result);
	}

	public function Begin() {
		return $this->Query("BEGIN", false);
	}

	public function Commit() {
		return $this->Query("COMMIT", false);
	}

	public function Rollback() {
		return $this->Query("ROLLBACK", false);
	}

	public function FreeResult() {
		$this->running = false;
		if($this->result !== false && $this->result !== true)
			mysql_free_result($this->result);
	}

	public function NextRecord() {
		if($this->running)
			if(($this->row = mysql_fetch_array($this->result)))
				return true;
		$this->running = false;
		return false;
	}

}

?>
